libs/libc/uio/lib_readv.c

```diff
@@ -68,56 +70,63 @@
 
 ssize_t readv(int fildes, FAR const struct iovec *iov, int iovcnt)
 {
-  ssize_t ntotal;
-  ssize_t nread;
-  size_t remaining;
   FAR uint8_t *buffer;
+  ssize_t nread;
+  size_t tocopy;
+  size_t total_size;
+  size_t offset;
   int i;
 
-  /* Process each entry in the struct iovec array */
-
-  for (i = 0, ntotal = 0; i < iovcnt; i++)
+  if (iovcnt == 0)
     {
-      /* Ignore zero-length reads */
-
-      if (iov[i].iov_len > 0)
-        {
-          buffer    = iov[i].iov_base;
-          remaining = iov[i].iov_len;
-
-          /* Read repeatedly as necessary to fill buffer */
-
-          do
-            {
-              /* NOTE:  read() is a cancellation point */
-
-              nread = read(fildes, buffer, remaining);
+      return 0;
+    }
 
-              /* Check for a read error */
+  if (iovcnt == 1)
+    {
+      return read(fildes, iov->iov_base, iov->iov_len);
+    }
 
-              if (nread < 0)
-                {
-                  return nread;
-                }
+  total_size = 0;
+  for (i = 0; i < iovcnt; i++)
+    {
+      total_size += iov[i].iov_len;
+    }
 
-              /* Check for an end-of-file condition */
+  if (total_size == 0)
+    {
+      return 0;
+    }
 
-              else if (nread == 0)
-                {
-                  return ntotal;
-                }
+  buffer = malloc(total_size);
```

Comment:
could we avoid malloc

Reply:
i can't think of any trivial ways.

Reply:
The only way to avoid malloc is to read directly into each iov and avoid the copy. In other words: iterate iov array. For each iov, if iov_len > 0, read() iov_len bytes into that iov. If any read() returns less than iov_len bytes, break from loop. Keep running counter of total bytes read.

I don't know if this is permissible under POSIX but it (1) avoids malloc, and (2) avoids copying. Even if there is added overhead from multiple calls to read(), it might be better than overhead of malloc(), free(), and copying.

Reply:
it doesn't work. please read the discussions in this PR.